0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,77 @@
+0. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call a backend to make sure the incoming funds are expected.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two sets of headers, like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Custom binary data is appended to the inner headers, after a blank line. This object is then encrypted and appended to the outer headers as binary data, after a blank line. Both inner and outer headers are parsed in the exact same way as HTTP headers.
+
+The outer headers come after a single status line, which must be `PSK/1.0`. The outer headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the inner headers.
+
+### Outer Request
+
+The outer headers and their data are formatted as below:
+
+```http
+PSK/1.0 PUBLIC
+Key: hmac-sha-256 fpwpAhlN588
+Header: data_everyone_can_see
+
+...
+```
+
+Everything stored in these headers is readable by all connectors between the sender and receiver. PSK payments use an [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) of the PSK details as a fulfillment, in order to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the inner PSK headers.
+
+Despite the fact that connectors can read the outer headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is end-to-end.
+
+The only headers in the outer headers should be the ones required for the receiver to decrypt the inner headers. In the above example, the decryption key for the outer headers' body is: `HMAC-SHA-256(data: "fpwpAhlN588", key: secret)`, where `secret` is the pre-shared secret key.
+
+| Header | Value |
+|--------|-------|
+| `Key` | _(Required)_ An algorithm name (`hmac-sha-256` is the only supported one for now) followed by a base64url-encoded nonce, used to generate the decryption key for the inner headers. |
+| ... | _(Optional)_ Additional headers. These can be read by any connectors. |
+
+
+### Inner Request
+
+The inner headers are structured like so:
+
+```http
+Expires-At: 2017-03-13T16:58:06.901Z
+Header-A: some_stuff
+Header-B: some_other_stuff
+Header-C: some_more_stuff
+Date: 2017-03-13T16:57:56.901Z
+Content-Type: application/json
+
+{"foo":"bar"}
+```
+
+Only the intended sender and receiver can decrypt and read the inner headers and their data. The inner headers contains an arbitrary message body preceded by the following headers:
```

Comment:
I'd suggest "encrypted headers" or "private headers".

So that would be:

* PSK version line: `PSK/1.0`
* Public headers
* Private headers
* Application-layer data (always private)